matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/EventRelationsAggregationTask.kt

```diff
@@ -374,4 +418,61 @@ internal class DefaultEventRelationsAggregationTask @Inject constructor(
             Timber.e("## Cannot find summary for key $reactionKey")
         }
     }
+
+    private fun handleVerification(realm: Realm, event: Event, roomId: String, isLocalEcho: Boolean, relatedEventId: String, userId: String) {
+        val eventSummary = EventAnnotationsSummaryEntity.where(realm, relatedEventId).findFirst()
+                ?: EventAnnotationsSummaryEntity.create(realm, roomId, relatedEventId).apply { this.roomId = roomId }
+
+        val verifSummary = eventSummary.referencesSummaryEntity
+                ?: ReferencesAggregatedSummaryEntity.create(realm, relatedEventId).also {
+                    eventSummary.referencesSummaryEntity = it
+                }
+
+        val txId = event.unsignedData?.transactionId
+
+        if (!isLocalEcho && verifSummary.sourceLocalEcho.contains(txId)) {
+            // ok it has already been handled
+        } else {
+            ContentMapper.map(verifSummary.content)?.toModel<ReferencesAggregatedContent>()
+            var data = ContentMapper.map(verifSummary.content)?.toModel<ReferencesAggregatedContent>()
+                    ?: ReferencesAggregatedContent(VerificationState.REQUEST.name)
+            // TODO ignore invalid messages? e.g a START after a CANCEL?
+            // i.e. never change state if already canceled/done
+            val newState = when (event.getClearType()) {
+                EventType.KEY_VERIFICATION_START  -> {
+                    VerificationState.WAITING
+                }
+                EventType.KEY_VERIFICATION_ACCEPT -> {
+                    VerificationState.WAITING
+                }
+                EventType.KEY_VERIFICATION_KEY    -> {
+                    VerificationState.WAITING
+                }
+                EventType.KEY_VERIFICATION_MAC    -> {
+                    VerificationState.WAITING
+                }
+                EventType.KEY_VERIFICATION_CANCEL -> {
+                    if (event.senderId == userId) {
+                        VerificationState.CANCELED_BY_ME
+                    } else VerificationState.CANCELED_BY_OTHER
+                }
+                EventType.KEY_VERIFICATION_DONE   -> {
+                    VerificationState.DONE
+                }
+                else                              -> VerificationState.REQUEST
+            }
+
+            data = data.copy(verificationSummary = newState.name)
+            verifSummary.content = ContentMapper.map(data.toContent())
+        }
+
+        if (isLocalEcho) {
+            verifSummary.sourceLocalEcho.add(event.eventId)
+        } else {
+            if (verifSummary.sourceLocalEcho.contains(txId)) {
```

Comment:
no need to test contain() before calling remove()

Reply:
Done